[Liaw Xin Yan] iP

Lemonsr opened the pull request:
# Ai-su
### An AI-powered (~~or not~~) Task Manager
> A link to somewhere [better](https://www.southparkstudios.com/random-episode) instead.

  👧 <( Okay, sheesh chill )
  👔\      💻 <( **Give me your tasks right now** 🔫 )
  👖     ====
 /  |  
 

## Current Progress
Week 4 tasks:
- [x] A-Gradle
- [x] A-Checkstyle
- [x] Level-10 GUI
- [x] A-Varargs

## 💡 Frequently Asked Questions (FAQ)
#### ❓ What _does_ Ai-su do?
- Manage tasks
- ...
- .. And that's it.

#### ❓ Should I use it?
Nah. There are better apps out there. But it was fun making this~

#### ❓ Can I view your source code?
Sure, but I write my code in [Chinese](https://en.wikipedia.org/wiki/Chinese_language). Here's the `main` method:
```java
 公共 静态 无效 Main（线[] args）会丢 AisuException {
    如果 （还在（this.fileName）） {
         Application.run(Aisu.class, args);
    } 不然 {
        丢 新 AisuException（“不行啊！”）；
    }
}
```
#### ❓ Can you tell me some good chinese pickup lines?
Yeah sure:
1. 你会弹吉他吗？不会！那你为什么能拨动我的心弦！
2. 你的脸上有点东西！有什么？有点帅~
3. 你能帮我洗一样东西吗？洗什么？喜欢我~

You can pm me for more! :)

Commit:
Add Gradle support

Lemonsr committed:
Add Level-1: Greet, Echo, Exit

Lemonsr committed:
Customize Duke (now Aisu)

Lemonsr committed:
Add Level-2: Add, List

Lemonsr committed:
Add Level-3: Mark as Done

Lemonsr committed:
Add Level-4: ToDo, Event, Deadline

Lemonsr committed:
Minor performance/code improvements:
- Convert from String concatenation to StringBuilder
- String checking with == to .equals()

Lemonsr committed:
Add Automated Text UI Testing

Lemonsr committed:
Add Level-5: Handle Errors

Lemonsr committed:
Add Level-6: Delete

Lemonsr committed:
Add Storage to save tasks in harddisk.

Lemonsr committed:
Add enums for task types

Lemonsr committed:
Integrate Storage with Aisu and add saving "done" status

Lemonsr committed:
add Level-8: Dates and Times (minimal)

Lemonsr committed:
Merge branch 'branch-Level-7'

# Conflicts:
#	src/main/java/Aisu.java (Fix merge conflict successfully)

Lemonsr committed:
Merge branch 'branch-Level-8'

# Conflicts:
#	src/main/java/Deadline.java (fix merge conflict successfully)
#	src/main/java/Tasklist.java (fix merge conflict successfully)

Lemonsr committed:
Add Parser class and Ui class for more OOP

Lemonsr committed:
Tasklist class: update switch case indentation to correct format

Lemonsr committed:
Tasklist class and Storage class: update switch case indentation to correct format

Lemonsr committed:
Merge branch 'master' of https://github.com/Lemonsr/ip

Lemonsr committed:
Add Command abstract class

Commands can be extracted into a class of objects to improve OOP usage.

Lemonsr committed:
Update Ui class and fix minor bugs

Lemonsr committed:
Aisu class: remove unused import

Lemonsr committed:
Add all java files to "aisu" package

Lemonsr committed:
Add aisu.command and aisu.task packages

Lemonsr committed:
Add JUnit test AisuTest
Tasklist class: extrapolate range check into checkError method

Range check is repeated in multiple methods. Extracting duplicated code
improves code reusability.

Lemonsr committed:
Add Junit tests for Storage class and Tasklist class

Lemonsr committed:
Add JUnit tests for Storage class and Tasklist class

Lemonsr committed:
Merge branch 'master' of https://github.com/Lemonsr/ip

Lemonsr committed:
Refactor (rename) "Tasklist" to "TaskList"

Lemonsr committed:
Create a JAR file for easy distribution

Lemonsr committed:
Add Javadocs to at least half of non-private classes/methods

Lemonsr committed:
Place line break to improve readability

Lemonsr committed:
Add a way to find tasks by searching for a keyword

Lemonsr committed:
Merge branch 'branch-A-JavaDoc'

Lemonsr committed:
Merge branch 'branch-A-CodingStandard'

Lemonsr committed:
Merge branch 'branch-Level-9'

# Conflicts:
#	src/main/java/aisu/Parser.java

Lemonsr committed:
FindCommand class: Add javadoc

Lemonsr committed:
Add more javadoc comments to Ui
Improve Ui formatting and code

Lemonsr committed:
Merge branch 'add-gradle-support'

Lemonsr committed:
Add Gradle support

Lemonsr committed:
Add dependency for javafx

Lemonsr committed:
Add Checkstyle

Lemonsr committed:
Attempt Checkstyle on ip

Lemonsr committed:
Add Launcher class and javafx improvements

Lemonsr committed:
Add base gui code

Lemonsr committed:
Rebuild scene using FXML

Lemonsr committed:
Fix MainWindow classNotFound error

Lemonsr committed:
Fix minor bug: Aisu response dialog not showing up correctly

Lemonsr committed:
Add Welcome message by Aisu when initialized

Lemonsr committed:
Add round profile image to improve UX

Lemonsr committed:
Add colour textbox and auto quit app with "bye"

Lemonsr committed:
Merge branch 'add-gui'

Lemonsr committed:
Merge conflict for add checkstyle and suppressions

Lemonsr committed:
Add Javadoc improvement based on checkstyle

Lemonsr committed:
Merge branch 'add-checkstyle'

Lemonsr committed:
Add missing Javadocs

Lemonsr committed:
Refactor all classes into their packages

Lemonsr committed:
Fix textbox resize issue

kleonang commented on build.gradle:
```diff
@@ -0,0 +1,63 @@
+    toolVersion = '8.29'
+}
+
+run{
```
```suggestion
run {
```
Consider having a space for more consistency within build.gradle?

kleonang commented on src/main/java/aisu/Aisu.java:
```diff
@@ -0,0 +1,93 @@
+public class Aisu {
+    private final Storage storage;
+    private TaskList tasklist;
+    private final Ui ui;
```
Ordering of fields can be standardised, e.g. private final fields before private fields as such:
```suggestion
    private final Storage storage;
    private final Ui ui;
    private TaskList tasklist;
```

kleonang commented on src/main/java/aisu/Aisu.java:
```diff
@@ -0,0 +1,93 @@
+
+
+/**
+ * A tasklist chatbot, named Aisu.
```
One empty line is sufficient before JavaDoc
```suggestion
import java.util.List;

/**
 * A tasklist chatbot, named Aisu.
```

kleonang commented on src/main/java/aisu/Aisu.java:
```diff
@@ -0,0 +1,93 @@
+    private TaskList tasklist;
+    private final Ui ui;
+
+    public Aisu(String dirPath, String fileName) {
```
Possible to add JavaDocs for public constructor

kleonang commented on src/main/java/aisu/command/AddCommand.java:
```diff
@@ -0,0 +1,62 @@
+import aisu.storage.Storage;
+import aisu.tasklist.TaskList;
+import aisu.ui.Ui;
+import aisu.task.Task;
```
It would be good to standardise imports lexicographically (e.g. aisu.a*** to aisu.z***)
```suggestion
import aisu.exception.AisuException;
import aisu.storage.Storage;
import aisu.task.Task;
import aisu.tasklist.TaskList;
import aisu.ui.Ui;
```

kleonang commented on src/main/java/aisu/command/AddCommand.java:
```diff
@@ -0,0 +1,62 @@
+    private final TaskList.TaskTypes taskType;
+    private final String input;
+
+    public AddCommand(TaskList.TaskTypes taskType, String input) {
```
Possible to add JavaDocs for public constructor

kleonang commented on src/main/java/aisu/command/DeleteCommand.java:
```diff
@@ -0,0 +1,47 @@
+import aisu.storage.Storage;
+import aisu.tasklist.TaskList;
+import aisu.ui.Ui;
+import aisu.task.Task;
```
Possible to order lexicographically as before
```suggestion
import aisu.exception.AisuException;
import aisu.storage.Storage;
import aisu.task.Task;
import aisu.tasklist.TaskList;
import aisu.ui.Ui;
```

kleonang commented on src/main/java/aisu/command/DeleteCommand.java:
```diff
@@ -0,0 +1,47 @@
+public class DeleteCommand extends Command {
+    private final int parseInt;
+
+    public DeleteCommand(int parseInt) {
```
Possible to include JavaDocs for public constructor

kleonang commented on src/main/java/aisu/command/ExitCommand.java:
```diff
@@ -0,0 +1,28 @@
+public class ExitCommand extends Command {
+
+    @Override
+    public void execute(TaskList tasklist, Storage storage, Ui ui) {
```
Possible to include JavaDocs for public method

kleonang commented on src/main/java/aisu/command/FindCommand.java:
```diff
@@ -0,0 +1,41 @@
+public class FindCommand extends Command {
+    private final String input;
+
+    public FindCommand(String input) {
```
Possible to include JavaDocs for public constructor

kleonang commented on src/main/java/aisu/command/MarkDoneCommand.java:
```diff
@@ -0,0 +1,45 @@
+public class MarkDoneCommand extends Command {
+    private final int parseInt;
+
+    public MarkDoneCommand(int parseInt) {
```
Possible to include JavaDocs for public constructor

kleonang commented on src/main/java/aisu/command/MarkDoneCommand.java:
```diff
@@ -0,0 +1,45 @@
+import aisu.storage.Storage;
+import aisu.tasklist.TaskList;
+import aisu.ui.Ui;
+import aisu.task.Task;
```
Lexicographical ordering would be good as before
```suggestion
import aisu.exception.AisuException;
import aisu.storage.Storage;
import aisu.task.Task;
import aisu.tasklist.TaskList;
import aisu.ui.Ui;
```

kleonang commented on src/main/java/aisu/exception/AisuException.java:
```diff
@@ -0,0 +1,23 @@
+ */
+public class AisuException extends Exception {
+
+    public AisuException(String message) {
```
Possible to include JavaDocs for public constructor

kleonang commented on src/main/java/aisu/parser/Parser.java:
```diff
@@ -0,0 +1,56 @@
+import aisu.command.FindCommand;
+import aisu.command.MarkDoneCommand;
+import aisu.command.ShowListCommand;
+import aisu.exception.AisuException;
```
Lexicographical ordering would be good as before
```suggestion
import aisu.command.AddCommand;
import aisu.command.Command;
import aisu.command.DeleteCommand;
import aisu.command.ExitCommand;
import aisu.command.FindCommand;
import aisu.command.MarkDoneCommand;
import aisu.command.ShowListCommand;
import aisu.exception.AisuException;
import aisu.tasklist.TaskList;
```

kleonang commented on src/main/java/aisu/storage/Storage.java:
```diff
@@ -0,0 +1,122 @@
+import aisu.task.Deadline;
+import aisu.task.Event;
+import aisu.task.Task;
+import aisu.task.Todo;
```
Lexicographical ordering would be good as before
```suggestion
import aisu.exception.AisuException;
import aisu.task.Deadline;
import aisu.task.Event;
import aisu.task.Task;
import aisu.task.Todo;
import aisu.tasklist.TaskList;
```

kleonang commented on src/main/java/aisu/storage/Storage.java:
```diff
@@ -0,0 +1,122 @@
+    private final String fileName;
+    private File file;
+
+    public Storage(String dirPath, String fileName) {
```
Possible to add JavaDocs for public constructor

kleonang commented on src/main/java/aisu/task/Deadline.java:
```diff
@@ -0,0 +1,49 @@
+public class Deadline extends Task {
+    private final LocalDate deadline;
+
+    public Deadline(String description, String deadline) throws AisuException {
```
Possible to add JavaDocs for public constructor

kleonang commented on src/main/java/aisu/task/Event.java:
```diff
@@ -0,0 +1,38 @@
+public class Event extends Task {
+    private final String eventPeriod;
+
+    public Event(String description, String eventPeriod) throws AisuException {
```
Possible to add JavaDocs for public constructor

kleonang commented on src/main/java/aisu/task/Task.java:
```diff
@@ -0,0 +1,53 @@
+    protected String description;
+    protected boolean isDone;
+
+    public Task(String description) throws AisuException {
```
Possible to add JavaDocs for public constructor

kleonang commented on src/main/java/aisu/task/Todo.java:
```diff
@@ -0,0 +1,36 @@
+ * @author Liaw Xin Yan
+ */
+public class Todo extends Task {
+    public Todo(String description) throws AisuException {
```
Possible to include JavaDocs for public constructor

kleonang commented on src/main/java/aisu/tasklist/TaskList.java:
```diff
@@ -0,0 +1,163 @@
+        T, D, E
+    }
+
+    public TaskList() {
+        this.list = new ArrayList<>();
+    }
+
+    public TaskList(List<Task> currList) {
+        this.list = currList;
+    }
```
Possible to add JavaDocs for these public constructors

kleonang commented on src/main/java/aisu/storage/Storage.java:
```diff
@@ -0,0 +1,122 @@
+import java.io.FileNotFoundException;
+import java.io.FileWriter;
+import java.io.IOException;
+
```
`import java.***` statements can be combined (no need new line separation) as in your `DialogueBox.java` class

kleonang commented on src/main/java/aisu/ui/Ui.java:
```diff
@@ -0,0 +1,151 @@
+package aisu.ui;
+
+import java.io.PrintStream;
+
+import java.util.Scanner;
```
`import java.***` statements can be combined (no need new line separation)

kleonang commented on src/main/java/aisu/ui/Ui.java:
```diff
@@ -0,0 +1,151 @@
+            + "|  | |  | |  |'->         \\      / \\  '-'(_ .'\n"
+            + "`--' `--' `--'             `-----'  `-----'   \n";
+    private static final int PADDING_LENGTH = 48;
+    private static String formattedInstructions;
+    private static final String GOODBYE_MESSAGE = "See you next time! :D";
```
`static final` fields can come before `static` fields to standardise ordering

kleonang commented on src/main/java/aisu/ui/Ui.java:
```diff
@@ -0,0 +1,151 @@
+    private final Scanner in;
+    private final PrintStream out;
+
+    public Ui() {
```
Possible to add JavaDocs for public constructor

kleonang commented on src/test/java/aisu/StorageTest.java:
```diff
@@ -0,0 +1,15 @@
+import aisu.exception.AisuException;
+import org.junit.jupiter.api.Test;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
```
Static imports can come before non-static imports

kleonang commented on src/test/java/aisu/TaskListTest.java:
```diff
@@ -0,0 +1,32 @@
+import java.util.ArrayList;
+import java.util.List;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
```
Static imports can come before non-static imports

kleonang reviewed (commented):
Coding standard is quite closely adhered to 👍. Main suggestions for changes are in inclusion of more JavaDocs and making import statements lexicographically ordered. 

Also this is quite nitpicky but there are 2 more potential improvements regarding coding standard, although they're more a matter of style:
1. Start regular comments with a capitalised first letter
2. Standardise spacing/alignment after JavaDoc tags (e.g. for the JavaDocs of the `execute` method in `AddCommand.java`, description of `@throws AisuException` is not aligned with those of the other `@param` tags)

Overall well done on the code! 🎉

Lemonsr committed:
Modify code based on PR review suggestions

There are several classes with Javadoc and import statement violations.

Following the suggestions given by the PR reviewer helps improve code
quality.

Writing Javadoc neatly will allow other reviewers to understand my code
better.

Lemonsr committed:
Update build.gradle to support GUI

Lemonsr committed:
Add Java Assertions

Lemonsr committed:
Refactor code: factor out magic numbers

Init instructions command uses magic numbers.

Magic numbers makes the code difficult to understand as there are no
explanations as to what they mean.

Variables describing what the number formulas mean is more user-friendly
and readable.

Let's factor out magic numbers to use variable containers.

Lemonsr committed:
Remove unnecessary whitespace at end of method bodies

Lemonsr committed:
Delete unused methods

Lemonsr committed:
Improve code quality

There are several methods that do not follow the code quality standards
specified in the textbook.

Refactoring the code improves code quality.

Lemonsr committed:
Add streams support

Lemonsr committed:
Merge pull request #2 from Lemonsr/branch-A-Streams

Add streams support

Lemonsr committed:
Merge pull request #1 from Lemonsr/branch-A-Assertions

Add Java Assertions

Lemonsr committed:
Merge pull request #3 from Lemonsr/branch-A-CodeQuality

Improve code quality

Lemonsr committed:
Minor Javadoc fixes and add author tags

Lemonsr committed:
Fix: Empty List does not print to GUI

Lemonsr committed:
Clean code and fix Ui bugs

Code has lots of methods for CLI in the Ui, though unnecessary now
with the existence of GUI.

Removing unnecessary methods can make the code simpler and more
readable for other programmers. It also makes it cleaner for debugging,
allows the code to be more expandable in the future, and reduces
unnecessary imports in other classes.

Lemonsr committed:
Add ability for users to tag items

Lemonsr committed:
Fix C-Tagging bugs and refactor code into Tag and TagList classes

- Tags are not printed if there are none.
- Tags are saved and loadable from storage

Lemonsr committed:
Add Ui file to docs

Lemonsr committed:
Update README.md

Lemonsr committed:
Set theme jekyll-theme-architect

Lemonsr committed:
Remove newline for displayed task when no tags exist

Lemonsr committed:
Fix parser uncaught exception bug

Lemonsr committed:
Fix minor checkstyle violations

Lemonsr committed:
Add latest release of Aisu chatbot

Lemonsr committed:
Delete libs

Lemonsr committed:
Add newest release of Aisu chatbot

Lemonsr committed:
Update README.md
